### ISSUE TITLE
where should layer norm be applied?

### ISSUE THREAD

**Issue by liangbright** [harvardnlp/annotated-transformer#13]
class SublayerConnection(nn.Module):

    def __init__(self, size, dropout):
        super(SublayerConnection, self).__init__()
        self.norm = LayerNorm(size)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, sublayer):
        "Apply residual connection to any sublayer with the same size."
        #return x + self.dropout(sublayer(self.norm(x)))
        return self.norm(x + self.dropout(sublayer(x)))

In the paper, it seems to be "return self.norm(x + self.dropout(sublayer(x)))". But it gives a bad result for the example "Train the simple copy task."

**Comment by gitfourteen** [comment]
Agree with you. 
Also confused about the **SublayerConnection()**. The text explanation quoted as below is contradicted the pytorch code.

'''
That is, the output of each sub-layer is LayerNorm(x+Sublayer(x)), where Sublayer(x) is the function implemented by the sub-layer itself. We apply dropout (cite) to the output of each sub-layer, **before** it is added to the sub-layer input and **normalized**.
'''

**Comment by daqingliu** [comment]
I find in the forward() commit:
"Note for code simplicity the norm is first as opposed to last."
But it is a mismatch with the paper.

**Comment by cbernet** [comment]
I'm necroing this thread a bit :-) But I agree with you guys.

We have these two classes : 

```python
class SublayerConnection(nn.Module):
    """
    A residual connection followed by a layer norm.
    Note for code simplicity the norm is first as opposed to last.
    """
    def __init__(self, size, dropout):
        super(SublayerConnection, self).__init__()
        self.norm = LayerNorm(size)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, sublayer):
        "Apply residual connection to any sublayer with the same size."
        return x + self.dropout(sublayer(self.norm(x)))  ##B
```

```python
class EncoderLayer(nn.Module):
    "Encoder is made up of self-attn and feed forward (defined below)"
    def __init__(self, size, self_attn, feed_forward, dropout):
        super(EncoderLayer, self).__init__()
        self.self_attn = self_attn
        self.feed_forward = feed_forward
        self.sublayer = clones(SublayerConnection(size, dropout), 2)
        self.size = size

    def forward(self, x, mask):
        "Follow Figure 1 (left) for connections."
        x = self.sublayer[0](x, lambda x: self.self_attn(x, x, x, mask))  ##A
        return self.sublayer[1](x, self.feed_forward)
```

If I follow line `##A` to go to `##B` I understand that : 

* normalization is applied BEFORE the self attention layer. If we consider only this layer, this might be equivalent. 
* norm is NOT applied to x (the residual connection), which is I think the real issue. 

I really don't see where x could be normalized. 

And if it's not, the residual connection does not make sense. The lack of normalization on x will make the residuals very large, which is killing their purpose. 

We could simply do the following, as pointed out by @liangbright 

```python
class SublayerConnection(nn.Module):
    """
    A residual connection followed by a layer norm.
    Note for code simplicity the norm is first as opposed to last.
    """
    def __init__(self, size, dropout):
        super(SublayerConnection, self).__init__()
        self.norm = LayerNorm(size)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, sublayer):
        "Apply residual connection to any sublayer with the same size."
        return self.norm(x + self.dropout(sublayer(x)))
```

and also remove the last norm here since with this change, the last normalization is carried on by the last `SubLayerConnection`

```python
class Encoder(nn.Module):
    "Core encoder is a stack of N layers"
    def __init__(self, layer, N):
        super(Encoder, self).__init__()
        self.layers = clones(layer, N)
        self.norm = LayerNorm(layer.size)
        
    def forward(self, x, mask):
        "Pass the input (and mask) through each layer in turn."
        for layer in self.layers:
            x = layer(x, mask)
        return self.norm(x)   ## remove
```

**Comment by srush** [comment]
Hi all, 

We are working on a new version of this writeup, and will make this point clear and why it was confusing. 

If you are interested in helping out come chat here: https://discord.gg/VUwHN8xZ

**Comment by XuRui314** [comment]
I have seen the new version, but i didn't find the explanation of this problem.  Does anyone see the updated writeup?

**Comment by XZJIsme** [comment]
> ``` python
> class SublayerConnection(nn.Module):
> def __init__(self, size, dropout):
>     super(SublayerConnection, self).__init__()
>     self.norm = LayerNorm(size)
>     self.dropout = nn.Dropout(dropout)
> 
> def forward(self, x, sublayer):
>     "Apply residual connection to any sublayer with the same size."
>     #return x + self.dropout(sublayer(self.norm(x)))
>     return self.norm(x + self.dropout(sublayer(x)))
> ```
> 
> In the paper, it seems to be "return self.norm(x + self.dropout(sublayer(x)))". But it gives a bad result for the example "Train the simple copy task."

Maybe it should be ```return (x + self.norm(self.dropout(sublayer(x))))```

**Comment by Silent-Zebra** [comment]
This paper might be relevant (explaining why the norm should be applied before rather than after): [On Layer Normalization in the Transformer Architecture](https://proceedings.mlr.press/v119/xiong20b/xiong20b.pdf).

**Comment by harumo11** [comment]
To beginners learning about transformer, just like me.

I'm not sure if here is the good place to write this, but for those who have the same question,
I will write down what I understand here.

## There are two type of Layer Normalization
- Post-Layer Normalization(Post-LN) : This method was used in Attention is All You Need
- Pre-Layer Normalization(Pre-LN) : This method was introduced in subsequent research

## The difference of Post-LN and Pre-LN
Please take a look this chart. That is quoted from [A Survey of Transformers](https://arxiv.org/pdf/2106.04554)
![image](https://storage.googleapis.com/zenn-user-upload/25f8cf4976f0-20231010.png)

- in Post-LN, Layer Normalization is put **after** each sub-layers such as Mult-Head Attention and Position-wise FFN.
- in Pre-LN, Layer Normalization is put **before** each sub-layers and encoder layer.

## Post-LN vs Pre-LN
- Final Performance:
Post-LN generally achieves superior final model performance compared to pre-LN, if the training process is successful.
- Training Stability:
Post-LN tends to exhibit greater instability during training compared to pre-LN, particularly the number of layers in the model increases.

## Details
In Attention is All You Need, Post-LN is used. After that paper, Pre-LN method is porposed. This method gives the transformer model more stability during learning. This method inserts layer normalization immediately before each attention or linear layer. And then, In gneral, in Pre-LN, an additional layer normalization is added after the final layer to maintain the norm scale of the final output.

I think annotated-transformer uses Pre-LN. That is not same as original paper. But it is more stable.

Thanks for reading.
